robot/impl/resource_manager_test.go

```diff
@@ -1850,3 +1852,72 @@ func managerForDummyRobot(t *testing.T, robot robot.Robot) *resourceManager {
 	}
 	return manager
 }
+
+// TestReconfigureParity validates that calling the synchronous and asynchronous version
+// of `Reconfigure` results in the same resource state on a local robot for all
+// combinations of an initial and updated configuration.
+func TestReconfigureParity(t *testing.T) {
+	files := []string{
+		"data/diff_config_deps1.json",
+		"data/diff_config_deps2.json",
+		"data/diff_config_deps3.json",
+		"data/diff_config_deps4.json",
+		"data/diff_config_deps5.json",
+		"data/diff_config_deps6.json",
+		"data/diff_config_deps7.json",
+		"data/diff_config_deps8.json",
+		"data/diff_config_deps9_good.json",
+		"data/diff_config_deps9_bad.json",
+		"data/diff_config_deps10.json",
+		"data/diff_config_deps11.json",
+		"data/diff_config_deps12.json",
+	}
+	logger := logging.NewTestLogger(t)
+	ctx := context.Background()
+
+	testReconfigureParity := func(t *testing.T, initCfg, updateCfg string) {
+		name := fmt.Sprintf("%s -> %s", initCfg, updateCfg)
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			// Configuration may mutate `*config.Config`, so we read it from
+			// file each time.
+			cfg := ConfigFromFile(t, initCfg)
+			r1 := setupLocalRobot(t, ctx, cfg, logger).(*localRobot)
+			cfg = ConfigFromFile(t, initCfg)
+			r2 := setupLocalRobot(t, ctx, cfg, logger).(*localRobot)
+
+			test.That(
+				t,
+				rdktestutils.NewResourceNameSet(r1.ResourceNames()...),
+				test.ShouldResemble,
+				rdktestutils.NewResourceNameSet(r2.ResourceNames()...),
+			)
+
+			cfg = ConfigFromFile(t, updateCfg)
+			r1.Reconfigure(ctx, cfg)
+			cfg = ConfigFromFile(t, updateCfg)
+			r2.ReconfigureSync(ctx, cfg)
+
+			test.That(
+				t,
+				rdktestutils.NewResourceNameSet(r1.ResourceNames()...),
+				test.ShouldResemble,
+				rdktestutils.NewResourceNameSet(r2.ResourceNames()...),
+			)
+		})
+	}
+
+	gen := combin.NewCombinationGenerator(len(files), 2)
+
+	pair := []int{0, 0}
+	for gen.Next() {
```

Comment:
[q] Does this test _every_ 2-combination? Like 80 tests (concurrently)? Seems fine, just wanted to check what this was doing.

Reply:
yep! since we run these tests in parallel I figured why not run through every combination

Reply:
also, each subtest is [named as `TestReconfigureParity/<init conf file> -> <final conf file>`](https://github.com/viamrobotics/rdk/pull/3987/files#diff-f60fc288e7fc9c4813d48d3c54b37fafcb3278e54f4b11704b4423b83b58f825R1879), so it should be pretty clear which combination failed despite the number of tests.